doc/designs/quic-design/quic-api.md

```diff
@@ -0,0 +1,153 @@
+be QUIC method objects (e.g. “QUIC_client_method'' or
+“QUIC_server_method”). The compatibility layer:
+
+1. Will call the generic API underneath, e.g. Common functions such as
+“SSL_read” and “SSL_write” would be redirected internally  to the
+equivalent generic connection/stream function as appropriate.
+2. For MVP it will only support 1 stream
+3. Should be possible (eventually) to move from the SSL API to the generic API
+e.g.
+
+    ```
+    OSSL_CONN *SSL_get_connection(SSL *ssl);
+
+    ```
+
+
+The new generic API could be internal only for MVP. We might expect it to have
+a context object type, a connection object type, a stream object type, etc.
+Since it is not TLS, DTLS or QUIC specific, we would expect a different prefix.
+
+Inevitably there will have to be mechanisms for making available protocol
+specific functionality as part of the generic API. An open question is how this
+would be done, e.g.
+
+* Exposing a protocol specific object
+* OSSL_PARAMs
+* Some other params style approach but with ability for complex objects (e.g.
+X509)
+
+A possible data model for this approach can be seen here ([whiteboard
+link](https://jamboard.google.com/d/1AM1g-9eUbGAJ3YR9cEVpH6mrpPHCxd-9DtJMG1hNPlk/edit?usp=meet_whiteboard)):
+
+
+![alt_text](images/image1.png "image_tooltip")
+
+
+Pros
+
+* All protocols look the same to an application. Straight forward to switch
+from one protocol to another
+
+Cons
+
+* Protocol differences will pollute the API
+* Inevitably there will be some features that will work differently (or won’t
+be present at all) between different protocols. Therefore applications swapping
+from one protocol to another may still have to have some protocol specific code
+
+
+## Candidate Solution: A completely new QUIC API with SSL compat layer
+
+This solution is similar to the previous solution except that we do not attempt
+to create a generic comms API and instead just implement a QUIC specific API.
+Applications wanting to do TLS would still use the existing “SSL” API, and
+applications wanting to do QUIC would use the new API.
+
+Additionally we would also implement an SSL “compatibility” layer that
+would convert SSL function calls to equivalent QUIC function as appropriate.
+The compatibility layer:
+
+1. Will call the QUIC API underneath, e.g. Common functions such as
+“SSL_read” and “SSL_write” would be redirected internally to the
+equivalent QUIC connection/stream function as appropriate.
+2. For MVP it will only support 1 stream
+3. Should be possible (eventually) to move from the SSL API to the QUIC API e.g.
+
+    ```
+    OSSL_QUIC_CONN *SSL_get_connection(SSL *ssl);
+
+    ```
+
+The new QUIC API could be internal only for MVP. We might expect it to have a
+context object type, a connection object type, a stream object type, etc.
+
+Pros
+
+* Simpler solution - we don’t have to fit into some bigger “generic”
+protocol concept
+
+Cons
+
+* Full QUIC will look very different to SSL
```

Comment:
I don't see this as a "con".  The actual protocols are different, and trying to hide those differences will eventually cause confusion, inefficiency, or both.